### PR TITLE
[objc] Significantly improve Objc Lexer and Parser + Tests

### PR DESCRIPTION
**Pull request by DJBen**
### Background
As an engineer of Snapchat, I build internal tools that parses existing objective-C api module files and translate them to Swift at scale. We found that the current syntax fails some cases and thus I fixed them, ran back tests and verified that current grammar can correctly parse 99.8% of our 7k+ .h files. The failed ones are due to Objective-C++ presence that is not considered to cover in this scope.

### Objc grammar
- Fixed issue of nested types with pointers and nullability specifiers may not get parsed correctly
  - Restructed syntax, so that pointer `*` is now a part of `typeSpecifier`.
- Add full support to parse API availability, Swift alias macros
- Fixed issue that [lightweight generics](https://medium.com/ios-os-x-development/generics-in-objective-c-8f54c9cfbce7) conformances in class interface cannot be parsed.
- Added 4 new test cases to cover the above mentioned cases.

### Others
- Corrected test path `examples` so that `maven clean test` can correctly find all the test files.
- Add testing section to objc's `README`

## Test & verify
`mvn clean test` in `objc/` dir.

### CODE REVIEW

File: objc/ObjectiveCParser.g4
Comment (kaby76) [inline objc/ObjectiveCParser.g4:425]: Useless parentheses: `( attributeSpecifier )*` => `attributeSpecifier*`.

File: objc/ObjectiveCParser.g4
Comment (kaby76) [inline objc/ObjectiveCParser.g4:511]: Group: `| typeSpecifierModifier* 'char' typeQualifier* | typeSpecifierModifier* 'short' typeQualifier* | typeSpecifierModifier* 'long' typeQualifier* ...` => `| typeSpecifierModifier* ('char' | 'short' | 'long' | ...)  typeQualifier*`

File: objc/ObjectiveCParser.g4
Comment (kaby76) [inline objc/ObjectiveCParser.g4:241]: `genericsSpecifierList` is unused!

Comment (DJBen) [reply objc/ObjectiveCParser.g4:241]: Good catch! ty

**Comment by DJBen** [comment]
@kaby76 Thank you Ken for your timely review. I've addressed the comments please let me know if they look good to you.

**Review by kaby76 — APPROVED** [review]

**Comment by DJBen** [comment]
@kaby76 Hi Ken it seems like I still need a maintainer's approval. Please let me know what should I do next. Thanks!

Screenshot 2024-08-08 at 10 18 02 AM

**Comment by kaby76** [comment]
> @kaby76 Hi Ken it seems like I still need a maintainer's approval before merge. Please let me know what should I do next. Thanks!

@teverett

**Pull request merged**
